src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,454 @@
+                        dataExistsAlready = true;
+                    }
+
+                    if (dataExistsAlready)
+                    {
+                        // The file was added but our collection hasn't resynced yet.
+                        // Set _certificates to null to force a resync.
+                        _certificates = null;
+                        return;
+                    }
+                }
+
+                string destinationFilename;
+                FileMode mode = FileMode.CreateNew;
+
+                if (existingFilename != null)
+                {
+                    destinationFilename = existingFilename;
+                    mode = FileMode.Create;
+                }
+                else if (findOpenSlot)
+                {
+                    destinationFilename = FindOpenSlot(thumbprint);
+                }
+                else
+                {
+                    destinationFilename = Path.Combine(_storePath, thumbprint + PfxExtension);
+                }
+
+                using (FileStream stream = new FileStream(destinationFilename, mode))
+                {
+                    byte[] pkcs12 = copy.Export(X509ContentType.Pkcs12);
+                    stream.Write(pkcs12, 0, pkcs12.Length);
+                }
+
+#if DEBUG
+                // Verify that we're creating files with u+rw and o-rw, g-rw.
+                const Interop.libc.Permissions requiredPermissions =
+                    Interop.libc.Permissions.S_IRUSR |
+                    Interop.libc.Permissions.S_IWUSR;
+
+                const Interop.libc.Permissions forbiddenPermissions =
+                    Interop.libc.Permissions.S_IROTH |
+                    Interop.libc.Permissions.S_IWOTH |
+                    Interop.libc.Permissions.S_IRGRP |
+                    Interop.libc.Permissions.S_IWGRP;
+
+                Interop.Sys.FileStatus stat;
+
+                Debug.Assert(Interop.Sys.Stat(destinationFilename, out stat) == 0);
+                Debug.Assert((stat.Mode & (int)requiredPermissions) != 0, "Created PFX has insufficient permissions to function");
+                Debug.Assert((stat.Mode & (int)forbiddenPermissions) == 0, "Created PFX has too broad of permissions");
+#endif
+            }
+
+            // Null out _certificates so the next call to get_Certificates causes a re-scan.
+            _certificates = null;
+        }
+
+        public void Remove(ICertificatePal certPal)
+        {
+            OpenSslX509CertificateReader cert = (OpenSslX509CertificateReader)certPal;
+
+            using (X509Certificate2 copy = new X509Certificate2(cert.DuplicateHandles()))
+            {
+                bool hadCandidates;
+                string currentFilename = FindExistingFilename(copy, _storePath, out hadCandidates);
+
+                if (currentFilename != null)
+                {
+                    if (_readOnly)
```

Comment:
Should we throw on attempted removal even if we can't find the cert in the store, e.g. moving this if up to the beginning of the method?  In Add, it looks like we throw in the similar case where we're adding a cert that already exists and thus it would be a nop.